builder/swig/csharp.bzl

```diff
@@ -0,0 +1,241 @@
+        includes = depset(
+            [h.dirname for h in ctx.attr.lib[CcInfo].compilation_context.headers.to_list()],
+            transitive = [lib_compilation_context.includes],
+        ),
+        local_defines = lib_compilation_context.local_defines,
+        quote_includes = lib_compilation_context.quote_includes,
+        system_includes = depset(
+            [],
+            transitive = [lib_compilation_context.system_includes],
+        )
+    )
+
+    return [
+        DefaultInfo(files = depset([wrap_cxx, wrap_csharp])),
+        OutputGroupInfo(
+            csharp_source = depset([wrap_csharp]),
+            cxx_source = depset([wrap_cxx]),
+        ),
+        CcInfo(
+            compilation_context = compilation_context,
+            linking_context = ctx.attr.lib[CcInfo].linking_context,
+        ),
+    ]
+
+
+swig_csharp_wrapper = rule(
+    implementation = _swig_csharp_wrapper_impl,
+    attrs = {
+        "lib": attr.label(
+            doc = "The C library for which to generate the wrapper",
+            providers = [CcInfo],
+            mandatory = True,
+        ),
+        "class_name": attr.string(
+            doc = "Optional override for the C# class name (default: same as target name)",
+        ),
+        "interface": attr.label(
+            doc = "Optional SWIG interface (.i) file",
+            allow_single_file = True,
+        ),
+        "includes": attr.label_list(
+            doc = "Additional SWIG files required for wrapper generation",
+            allow_files = True,
+        ),
+        "namespace": attr.string(
+            doc = "C# namespace where to put the sources",
+            mandatory = True,
+        ),
+        "enable_cxx": attr.bool(
+            doc = "Enable SWIG C++ processing (default: False)",
+            default = False,
+        ),
+        "extra_args": attr.string_list(
+            doc = "Extra arguments to be passed to SWIG",
+        ),
+        "_swig": attr.label(
+            default = Label("@swig//:swig"),
+            allow_single_file = True,
+            executable = True,
+            cfg = "host",
+        ),
+    },
+)
+
+
+def _csharp_native_library(ctx):
+    return [
+        DefaultInfo(
+            files = depset([ctx.file.native_lib]),
+            runfiles = ctx.runfiles(files = [ctx.file.native_lib])
+        ),
+        DotnetAssemblyCompileInfo(
+            name = "{}".format(ctx.file.native_lib.basename),
+            internals_visible_to = ctx.attr.name,
+            refs = [],
+            irefs = [],
+            analyzers = [],
+            compile_data = [ctx.file.native_lib],
+            exports = [],
+        ),
+        DotnetAssemblyRuntimeInfo(
+            name = "{}".format(ctx.file.native_lib.basename),
+            version = "",
+            deps = depset(),
+            nuget_info = [],
+            libs = [],
+            native = [ctx.file.native_lib],
+            direct_deps_depsjson_fragment = [],
+        )]
```

Comment:
* How many of these empty args are defaulted?
* `"{}".format(...)` does nothing, `File.basename` is already a `string`

Reply:
* Not a single one, I was inserting it while getting errors.
* I know, but it looked better for me to have all the rules names in a similar format with .format. I can remove it if it disturbs a reader too much.

Reply:
Approved irl